File: binding/src/multimedia/libqaudioformat.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qaudioformat.html
pub const qaudioformat = struct {

    /// New constructs a new QAudioFormat object.
    pub fn New(other: ?*anyopaque) QtC.QAudioFormat {
        return qtc.QAudioFormat_new(@ptrCast(other));
    }


    /// New2 constructs a new QAudioFormat object.
    pub fn New2(other: ?*anyopaque) QtC.QAudioFormat {
        return qtc.QAudioFormat_new2(@ptrCast(other));
    }


    /// New3 constructs a new QAudioFormat object.
    pub fn New3() QtC.QAudioFormat {
        return qtc.QAudioFormat_new3();
    }


    /// New4 constructs a new QAudioFormat object.
    pub fn New4(param1: ?*anyopaque) QtC.QAudioFormat {
        return qtc.QAudioFormat_new4(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QAudioFormat_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QAudioFormat_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QAudioFormat_IsValid(@ptrCast(self));
    }

    pub fn SetSampleRate(self: ?*anyopaque, sampleRate: i32) void {
        qtc.QAudioFormat_SetSampleRate(@ptrCast(self), sampleRate);
    }

    pub fn SampleRate(self: ?*anyopaque, ) i32 {
        return qtc.QAudioFormat_SampleRate(@ptrCast(self));
    }

    pub fn SetChannelConfig(self: ?*anyopaque, config: u32) void {
        qtc.QAudioFormat_SetChannelConfig(@ptrCast(self), @intCast(config));
    }

    pub fn ChannelConfig(self: ?*anyopaque, ) u32 {
        return qtc.QAudioFormat_ChannelConfig(@ptrCast(self));
    }

    pub fn SetChannelCount(self: ?*anyopaque, channelCount: i32) void {
        qtc.QAudioFormat_SetChannelCount(@ptrCast(self), channelCount);
    }

    pub fn ChannelCount(self: ?*anyopaque, ) i32 {
        return qtc.QAudioFormat_ChannelCount(@ptrCast(self));
    }

    pub fn ChannelOffset(self: ?*anyopaque, channel: i32) i32 {
        return qtc.QAudioFormat_ChannelOffset(@ptrCast(self), @intCast(channel));
    }

    pub fn SetSampleFormat(self: ?*anyopaque, f: u16) void {
        qtc.QAudioFormat_SetSampleFormat(@ptrCast(self), @intCast(f));
    }

    pub fn SampleFormat(self: ?*anyopaque, ) u16 {
        return qtc.QAudioFormat_SampleFormat(@ptrCast(self));
    }

    pub fn BytesForDuration(self: ?*anyopaque, microseconds: i64) i32 {
        return qtc.QAudioFormat_BytesForDuration(@ptrCast(self), microseconds);
    }

    pub fn DurationForBytes(self: ?*anyopaque, byteCount: i32) i64 {
        return qtc.QAudioFormat_DurationForBytes(@ptrCast(self), byteCount);
    }

    pub fn BytesForFrames(self: ?*anyopaque, frameCount: i32) i32 {
        return qtc.QAudioFormat_BytesForFrames(@ptrCast(self), frameCount);
    }

    pub fn FramesForBytes(self: ?*anyopaque, byteCount: i32) i32 {
        return qtc.QAudioFormat_FramesForBytes(@ptrCast(self), byteCount);
    }

    pub fn FramesForDuration(self: ?*anyopaque, microseconds: i64) i32 {
        return qtc.QAudioFormat_FramesForDuration(@ptrCast(self), microseconds);
    }

    pub fn DurationForFrames(self: ?*anyopaque, frameCount: i32) i64 {
        return qtc.QAudioFormat_DurationForFrames(@ptrCast(self), frameCount);
    }

    pub fn BytesPerFrame(self: ?*anyopaque, ) i32 {
        return qtc.QAudioFormat_BytesPerFrame(@ptrCast(self));
    }

    pub fn BytesPerSample(self: ?*anyopaque, ) i32 {
        return qtc.QAudioFormat_BytesPerSample(@ptrCast(self));
    }

    pub fn NormalizedSampleValue(self: ?*anyopaque, sample: ?*anyopaque) f32 {
        return qtc.QAudioFormat_NormalizedSampleValue(@ptrCast(self), @ptrCast(sample));
    }

    pub fn DefaultChannelConfigForChannelCount(channelCount: i32) u32 {
        return qtc.QAudioFormat_DefaultChannelConfigForChannelCount(channelCount);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAudioFormat_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/audioformat.html#types
pub const enums = struct {
    pub const SampleFormat = enum {
        pub const Unknown: u16 = 0;
        pub const UInt8: u16 = 1;
        pub const Int16: u16 = 2;
        pub const Int32: u16 = 3;
        pub const Float: u16 = 4;
        pub const NSampleFormats: u16 = 5;
    };

    pub const AudioChannelPosition = enum {
        pub const UnknownPosition: i32 = 0;
        pub const FrontLeft: i32 = 1;
        pub const FrontRight: i32 = 2;
        pub const FrontCenter: i32 = 3;
        pub const LFE: i32 = 4;
        pub const BackLeft: i32 = 5;
        pub const BackRight: i32 = 6;
        pub const FrontLeftOfCenter: i32 = 7;
        pub const FrontRightOfCenter: i32 = 8;
        pub const BackCenter: i32 = 9;
        pub const SideLeft: i32 = 10;
        pub const SideRight: i32 = 11;
        pub const TopCenter: i32 = 12;
        pub const TopFrontLeft: i32 = 13;
        pub const TopFrontCenter: i32 = 14;
        pub const TopFrontRight: i32 = 15;
        pub const TopBackLeft: i32 = 16;
        pub const TopBackCenter: i32 = 17;
        pub const TopBackRight: i32 = 18;
        pub const LFE2: i32 = 19;
        pub const TopSideLeft: i32 = 20;
        pub const TopSideRight: i32 = 21;
        pub const BottomFrontCenter: i32 = 22;
        pub const BottomFrontLeft: i32 = 23;
        pub const BottomFrontRight: i32 = 24;
    };

    pub const ChannelConfig = enum {
        pub const ChannelConfigUnknown: u32 = 0;
        pub const ChannelConfigMono: u32 = 8;
        pub const ChannelConfigStereo: u32 = 6;
        pub const ChannelConfig2Dot1: u32 = 22;
        pub const ChannelConfig3Dot0: u32 = 14;
        pub const ChannelConfig3Dot1: u32 = 30;
        pub const ChannelConfigSurround5Dot0: u32 = 110;
        pub const ChannelConfigSurround5Dot1: u32 = 126;
        pub const ChannelConfigSurround7Dot0: u32 = 3182;
        pub const ChannelConfigSurround7Dot1: u32 = 3198;
    };

};

File: binding/src/multimedia/libqaudioformat.cpp
#include <QAudioFormat>
#include <qaudioformat.h>
#include "libqaudioformat.h"
#include "libqaudioformat.hxx"

QAudioFormat* QAudioFormat_new(const QAudioFormat* other) {
	 return new QAudioFormat(*other);
}

QAudioFormat* QAudioFormat_new2(QAudioFormat* other) {
	 return new QAudioFormat(*other);
}

QAudioFormat* QAudioFormat_new3() {
	 return new QAudioFormat();
}

QAudioFormat* QAudioFormat_new4(const QAudioFormat* param1) {
	 return new QAudioFormat(*param1);
}

void QAudioFormat_CopyAssign(QAudioFormat* self, QAudioFormat* other) {
    *self = *other;
}

void QAudioFormat_MoveAssign(QAudioFormat* self, QAudioFormat* other) {
    *self = std::move(*other);
}

bool QAudioFormat_IsValid(const QAudioFormat* self) {
	return self->isValid();
}

void QAudioFormat_SetSampleRate(QAudioFormat* self, int sampleRate) {
	self->setSampleRate(sampleRate);
}

int QAudioFormat_SampleRate(const QAudioFormat* self) {
	return self->sampleRate();
}

void QAudioFormat_SetChannelConfig(QAudioFormat* self, quint32 config) {
	self->setChannelConfig(static_cast<QAudioFormat::ChannelConfig>(config));
}

quint32 QAudioFormat_ChannelConfig(const QAudioFormat* self) {
	return self->channelConfig();
}

void QAudioFormat_SetChannelCount(QAudioFormat* self, int channelCount) {
	self->setChannelCount(channelCount);
}

int QAudioFormat_ChannelCount(const QAudioFormat* self) {
	return self->channelCount();
}

int QAudioFormat_ChannelOffset(const QAudioFormat* self, int channel) {
	return self->channelOffset(static_cast<QAudioFormat::AudioChannelPosition>(channel));
}

void QAudioFormat_SetSampleFormat(QAudioFormat* self, quint16 f) {
	self->setSampleFormat(static_cast<QAudioFormat::SampleFormat>(f));
}

quint16 QAudioFormat_SampleFormat(const QAudioFormat* self) {
	return self->sampleFormat();
}

int QAudioFormat_BytesForDuration(const QAudioFormat* self, long long microseconds) {
	return self->bytesForDuration(microseconds);
}

long long QAudioFormat_DurationForBytes(const QAudioFormat* self, int byteCount) {
	return self->durationForBytes(byteCount);
}

int QAudioFormat_BytesForFrames(const QAudioFormat* self, int frameCount) {
	return self->bytesForFrames(frameCount);
}

int QAudioFormat_FramesForBytes(const QAudioFormat* self, int byteCount) {
	return self->framesForBytes(byteCount);
}

int QAudioFormat_FramesForDuration(const QAudioFormat* self, long long microseconds) {
	return self->framesForDuration(microseconds);
}

long long QAudioFormat_DurationForFrames(const QAudioFormat* self, int frameCount) {
	return self->durationForFrames(frameCount);
}

int QAudioFormat_BytesPerFrame(const QAudioFormat* self) {
	return self->bytesPerFrame();
}

int QAudioFormat_BytesPerSample(const QAudioFormat* self) {
	return self->bytesPerSample();
}

float QAudioFormat_NormalizedSampleValue(const QAudioFormat* self, const void* sample) {
	return self->normalizedSampleValue(sample);
}

quint32 QAudioFormat_DefaultChannelConfigForChannelCount(int channelCount) {
	return QAudioFormat::defaultChannelConfigForChannelCount(channelCount);
}

void QAudioFormat_Delete(QAudioFormat* self) {
    delete self;
}

